source/common/upstream/upstream_impl.cc

```diff
@@ -672,16 +690,46 @@ class FactoryContextImpl : public Server::Configuration::CommonFactoryContext {
   Api::Api& api_;
 };
 
+const std::shared_ptr<const ClusterInfoImpl::HttpProtocolOptionsConfigImpl> createOptions(
+    const envoy::config::cluster::v3::Cluster& config,
+    const std::shared_ptr<const ClusterInfoImpl::HttpProtocolOptionsConfigImpl>&& options) {
```

Comment:
This is a bit beyond my C++ foo, but passing a const shared_ptr by move and then moving it seems wrong / not doing what you think it's doing? Should it be passed and returned as non-const?